src/Persistence/MassTransit.DynamoDbIntegration/DynamoDbIntegration/Saga/DynamoDbDatabaseContext.cs

```diff
@@ -46,37 +45,41 @@ public async Task<TSaga> Load(Guid correlationId)
         public async Task Update(SagaConsumeContext<TSaga> context)
         {
             var instance = context.Saga;
-            var updateLock = await Lock(instance, context.CancellationToken).ConfigureAwait(false);
             try
             {
-                instance.Version++;
-
-                var existingInstance = await Load(instance.CorrelationId).ConfigureAwait(false);
-                if (existingInstance.Version >= instance.Version)
-                    throw new DynamoDbSagaConcurrencyException("Saga version conflict", typeof(TSaga), instance.CorrelationId);
-
-                await Save(instance).ConfigureAwait(false);
+                var toUpdateSaga = new DynamoDbSaga
+                {
+                    CorrelationId = _options.FormatSagaKey(instance.CorrelationId),
+                    VersionNumber = ++instance.Version,
+                    Properties = JsonSerializer.Serialize(instance, SystemTextJsonMessageSerializer.Options),
+                    ExpirationEpochSeconds = _options.Expiration.HasValue
+                        ? DateTimeOffset.UtcNow.Add(_options.Expiration.Value).ToUnixTimeSeconds()
+                        : (long?)null
+                };
+
+                await _database.GetTargetTable<DynamoDbSaga>(_options.DefaultConfig())
+                    .UpdateItemAsync(toUpdateSaga.ToDocument(), new Primitive(toUpdateSaga.CorrelationId), new Primitive(DynamoDbSaga.DefaultEntityType),
+                        BuildVersionNumberCondition(toUpdateSaga.VersionNumber - 1));
             }
-            catch (Exception exception)
+            catch (ConditionalCheckFailedException)
             {
-                throw new SagaException("Saga update failed", typeof(TSaga), instance.CorrelationId, exception);
+                throw new DynamoDbSagaConcurrencyException("Saga version conflict", typeof(TSaga), instance.CorrelationId);
             }
-            finally
+            catch (Exception exception)
             {
-                updateLock?.DisposeAsync().ConfigureAwait(false);
+                throw new SagaException("Saga update failed", typeof(TSaga), instance.CorrelationId, exception);
             }
         }
 
         public Task Delete(SagaConsumeContext<TSaga> context)
         {
-            return _database.DeleteAsync(new DynamoDbSaga { CorrelationId = _options.FormatSagaKey(context.Saga.CorrelationId) },
+            return _database.DeleteAsync(new DynamoDbSaga {CorrelationId = _options.FormatSagaKey(context.Saga.CorrelationId)},
```

Comment:
Delete should have the same conditional check, but only delete if the version matches the loaded version (not incremented.